### PR TITLE
Several minor fixes to proxy support in ClickHouse

### PR DESCRIPTION
**Pull request by arthurpassos**
1. Fix wrong argument (http method, e.g GET) being passed to `getProxyConfiguration` as protocol (http or https). [Reference](https://github.com/ClickHouse/ClickHouse/blob/652e2e052517b833a9d96c84085be8be9d2de0dd/src/Storages/StorageURL.cpp#L460)
2. Fix missing proxy argument while building `BuilderRWBufferFromHTTP`. [Reference](https://github.com/ClickHouse/ClickHouse/blob/652e2e052517b833a9d96c84085be8be9d2de0dd/src/Storages/StorageURL.cpp#L1322)
3. Fix wrong default for `disable_tunneling_for_https_requests_over_http_proxy_` on RemoteProxyConfigurationResolver. [Reference](https://github.com/ClickHouse/ClickHouse/blob/652e2e052517b833a9d96c84085be8be9d2de0dd/src/Common/RemoteProxyConfigurationResolver.h#L31).
4. Fix stale proxy protocol being used to decide if tunneling should be used. [Reference](https://github.com/ClickHouse/ClickHouse/blob/652e2e052517b833a9d96c84085be8be9d2de0dd/src/Common/RemoteProxyConfigurationResolver.cpp#L100).
5. Adjust tests so unexpected http methods in proxy logs are treated as errors. This helps assert tunneling is not on when it should be off.
6. Fix remote resolver cache not being used
7. Proxy configuration resolve was not being called inside try-catch block, making it non-retryable.

-- 1: Corner case, affects HTTPS requests over HTTP proxy. It caused `StorageURL` to misconfigure proxy tunneling and disable it.
-- 2:  `IStorageURLBase::tryGetLastModificationTime` request would not go through the proxy.
-- 3: Corner case, affects HTTPS requests over HTTP proxy when the proxy remote resolver is used. Caused the tunneling to be disabled for HTTPS requests over HTTP proxy. Similar to -- 1 -- .
-- 4: Same as 1 and 3.
-- 6: Multiple calls to remote resolver
-- 7: remote proxy configuration resolver not being retryable.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Several minor corner case fixes to proxy support & tunneling

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>Modify your CI run</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  All with Azure
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

<details>

### CODE REVIEW

File: src/Storages/StorageURL.cpp [hunk 1]
```diff
@@ -1327,6 +1328,7 @@ std::optional<time_t> IStorageURLBase::tryGetLastModificationTime(
                    .withBufSize(settings.max_read_buffer_size)
                    .withRedirects(settings.max_http_get_redirects)
                    .withHeaders(headers)
+                   .withProxy(proxy_config)
```
Comment (arthurpassos) [inline src/Storages/StorageURL.cpp:1331]: cc @CheSema I believe you forgot to include proxy configuration https://github.com/ClickHouse/ClickHouse/pull/58845

File: src/Storages/StorageURL.cpp [hunk 2]
```diff
@@ -457,7 +457,7 @@ std::pair<Poco::URI, std::unique_ptr<ReadWriteBufferFromHTTP>> StorageURLSource:
 
         const auto settings = context_->getSettings();
 
-        auto proxy_config = getProxyConfiguration(http_method);
```
Comment (arthurpassos) [inline src/Storages/StorageURL.cpp:460]: This one is on me. In some places, the protocol is called scheme and even method. This caused some confusion and it somehow got away untested.

**Comment by alexey-milovidov** [comment]
A test?

File: src/Common/RemoteProxyConfigurationResolver.h [hunk 3]
```diff
@@ -10,6 +10,19 @@
 namespace DB
 {
 
+struct ConnectionTimeouts;
+
+struct RemoteProxyHostFetcher
```
Comment (arthurpassos) [inline src/Common/RemoteProxyConfigurationResolver.h:15]: Created this interface so the HTTP response can be mocked and some level of unit testing can be achieved

**Comment by arthurpassos** [comment]
@alexey-milovidov You can find the full details of this fix in the PR description, but mainly there are three fixes and they are not easy to test. Below you can find a summary.

1. **StorageURL - HTTP method (e.g, GET) being passed to a function that expects the protocol (e.g, http or https), which had the side effect of disabling tunneling.** _Tunneling is only used for HTTPS requests over HTTP proxies. The nginx server that is used for proxying in integration tests does not properly handle the HTTP CONNECT method and I believe it does not support it. Therefore, it can't be tested properly (I see the incoming CONNECT requests in the nginx log, but the tunnel is not formed and the ClickHouse requests timeout)_. 
2. **StorageURL - Proxy config not being passed to HTTP request builder when `StorageURLBase::tryGetLastModificationTime` is called.** _Also tricky to test. AFAIK, there is no way to invoke only this function. Invoking any regular communication with URLs would also call the other URL functions which make proper use of the proxy server, which fills the proxy logs with HTTP requests even though it `StorageURLBase::tryGetLastModificationTime` did not go through it_
3. **Incorrect originalRequestProtocol being used to decide if tunneling should be on or off on `RemoteProxyConfigurationResolver`.** _I have added a unit tests for that_.

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit a64cf57375950a386e29a1dfe181270bb3ce9a12 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63427/a64cf57375950a386e29a1dfe181270bb3ce9a12/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Common/RemoteProxyConfigurationResolver.cpp [hunk 4]
```diff
@@ -16,20 +16,63 @@ namespace ErrorCodes
     extern const int BAD_ARGUMENTS;
 }
 
+std::string RemoteProxyHostFetcherImpl::fetch(const Poco::URI & endpoint, const ConnectionTimeouts & timeouts) const
+{
+    /// It should be just empty GET request.
+    Poco::Net::HTTPRequest request(Poco::Net::HTTPRequest::HTTP_GET, endpoint.getPath(), Poco::Net::HTTPRequest::HTTP_1_1);
+
+    const auto & host = endpoint.getHost();
+    auto resolved_hosts = DNSResolver::instance().resolveHostAll(host);
```
Comment (CheSema) [inline src/Common/RemoteProxyConfigurationResolver.cpp:25]: Why do you use here `DNSResolver`?

File: src/Common/RemoteProxyConfigurationResolver.cpp [hunk 5]
```diff
@@ -16,20 +16,63 @@ namespace ErrorCodes
...
+
+    HTTPSessionPtr session;
+
+    for (size_t i = 0; i < resolved_hosts.size(); ++i)
+    {
+        auto resolved_endpoint = endpoint;
+        resolved_endpoint.setHost(resolved_hosts[i].toString());
+        session = makeHTTPSession(HTTPConnectionGroupType::HTTP, resolved_endpoint, timeouts);
```
[the remark this answers is not in the document]
Comment (CheSema) [reply src/Common/RemoteProxyConfigurationResolver.cpp:33]: makeHTTPSession could trow exception if it fails on connection establishing. That should be under `try catch`.

File: src/Common/RemoteProxyConfigurationResolver.cpp [hunk 6]
```diff
@@ -16,20 +16,63 @@ namespace ErrorCodes
...
+{
+    /// It should be just empty GET request.
+    Poco::Net::HTTPRequest request(Poco::Net::HTTPRequest::HTTP_GET, endpoint.getPath(), Poco::Net::HTTPRequest::HTTP_1_1);
+
+    const auto & host = endpoint.getHost();
+    auto resolved_hosts = DNSResolver::instance().resolveHostAll(host);
+
+    HTTPSessionPtr session;
+
+    for (size_t i = 0; i < resolved_hosts.size(); ++i)
```
Comment (CheSema) [inline src/Common/RemoteProxyConfigurationResolver.cpp:29]: Do you intended to do retries? Why is it depend on IP's count in resolve result?
I think it is better use here some our common class which does requests with good retries.

File: src/Common/RemoteProxyConfigurationResolver.cpp [hunk 7]
```diff
@@ -16,20 +16,63 @@ namespace ErrorCodes
...
+
+    Poco::Net::HTTPResponse response;
+    auto & response_body_stream = session->receiveResponse(response);
+
+    if (response.getStatus() != Poco::Net::HTTPResponse::HTTP_OK)
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Proxy resolver returned not OK status: {}", response.getReason());
+
+    String proxy_host;
+    /// Read proxy host as string from response body.
+    Poco::StreamCopier::copyToString(response_body_stream, proxy_host);
```
Comment (CheSema) [inline src/Common/RemoteProxyConfigurationResolver.cpp:55]: that read also could throw exception.

File: src/Common/RemoteProxyConfigurationResolver.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (arthurpassos) [reply src/Common/RemoteProxyConfigurationResolver.cpp:33]: This comment answers your other questions. TLDR: not sure. This code has been extracted twice from its original implementation and it was already like this. I don't see any obvious reason as to why I need to manually resolve the hostnames and even loop over the ips.

I could try to simplify the logic if you want, no side effects come to mind but I can't guarantee. Simplification would be something like: simply call `makeHTTPSession` with the URI.

What do you think?

File: src/Common/RemoteProxyConfigurationResolver.cpp (hunk 7 above)
Comment (arthurpassos) [reply src/Common/RemoteProxyConfigurationResolver.cpp:55]: Do we need to catch it tho? I think it's ok to throw the exception in case fetching the proxy host fails. I wouldn't want to have a proxy server configured that is bypassed when it fails. What do you think?

File: src/Common/RemoteProxyConfigurationResolver.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (CheSema) [reply src/Common/RemoteProxyConfigurationResolver.cpp]: > simply call makeHTTPSession with the URI.

That is good. That is what I expected to see here.

File: src/Common/RemoteProxyConfigurationResolver.cpp (hunk 7 above)
Comment (CheSema) [reply src/Common/RemoteProxyConfigurationResolver.cpp:36]: We need to answer to the question: "who is responsible for making retries here?".
Do AWS SDK retry proxy resolve call? If so, than it is enough and we should do any retry here, just pass the error up.
If anyone is retry that proxy resolve calls, that we have to do it here.

There is a possibility to get just network error in each call which has been mentioned above. That errors are not fatal, they have to be retried.

File: src/Common/RemoteProxyConfigurationResolver.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (CheSema) [reply src/Common/RemoteProxyConfigurationResolver.cpp]: > it was already like this.

I understand that it was like this. It was bad from my point of view. Lets change it as well. We don't have to preserve it in its original shape for the future.

File: src/Common/RemoteProxyConfigurationResolver.cpp (hunk 7 above)
Comment (arthurpassos) [reply src/Common/RemoteProxyConfigurationResolver.cpp:36]: Ok, AFAIK Poco does not implement retry logic by default. I can implement a simple one with a fixed number of retries.

Comment (CheSema) [reply src/Common/RemoteProxyConfigurationResolver.cpp:36]: Try to use ReadWriteBufferFromHTTP. It does retries.

Comment (arthurpassos) [reply src/Common/RemoteProxyConfigurationResolver.cpp]: Done

Comment (arthurpassos) [reply src/Common/RemoteProxyConfigurationResolver.cpp]: Is there a way I can check for the http status code? I can't find an API that allows me to do that

Comment (CheSema) [reply src/Common/RemoteProxyConfigurationResolver.cpp]: ReadWriteBufferFromHTTP checks it. If it bad, the request is either retried or an excpection is thrown.

File: src/Common/RemoteProxyConfigurationResolver.cpp [hunk 8]
```diff
@@ -7,29 +7,44 @@
...
 
 namespace DB
 {
 
-namespace ErrorCodes
+std::string RemoteProxyHostFetcherImpl::fetch(const Poco::URI & endpoint, const ConnectionTimeouts & timeouts) const
 {
-    extern const int BAD_ARGUMENTS;
+    /// It should be just empty GET request.
+    Poco::Net::HTTPRequest request(Poco::Net::HTTPRequest::HTTP_GET, endpoint.getPath(), Poco::Net::HTTPRequest::HTTP_1_1);
```
Comment (CheSema) [inline src/Common/RemoteProxyConfigurationResolver.cpp:18]: unused local variable.

File: src/Common/RemoteProxyConfigurationResolver.cpp (hunk 7 above)
Comment (arthurpassos) [reply src/Common/RemoteProxyConfigurationResolver.cpp]: Ok, there has been a small confusion. When you asked me if AWS takes cares of retries for proxy calls, I thought you were asking about the client we use to make the proxy calls, that's why I said no retries were in place.

AWS indeed retries this. If I add retry logic to proxy retrieval, we get retries of retries - we should avoid it. Even if AWS did not retry, I don't think we should implement retries here. I think this should be up to the caller, otherwise it might be "uncontrollable".

Therefore, I suggest we just keep it simple: either `makeHTTPSession` or `ReadWriteBufferFromHTTP` with max_http_Retry=1

File: src/Common/RemoteProxyConfigurationResolver.cpp [hunk 9]
```diff
@@ -7,29 +7,46 @@
 #include <Poco/Net/HTTPResponse.h>
 #include <Common/logger_useful.h>
 #include <Common/DNSResolver.h>
+#include <IO/ReadWriteBufferFromHTTP.h>
 
 namespace DB
 {
 
-namespace ErrorCodes
+std::string RemoteProxyHostFetcherImpl::fetch(const Poco::URI & endpoint, const ConnectionTimeouts & timeouts) const
```
Comment (CheSema) [inline src/Common/RemoteProxyConfigurationResolver.cpp:15]:
```
May 17 09:38:22 FAILED: src/CMakeFiles/clickhouse_common_io.dir/Common/RemoteProxyConfigurationResolver.cpp.o 
May 17 09:38:22    23 |     std::string fetch(const Poco::URI & uri, const ConnectionTimeouts & timeouts) const override;
May 17 09:38:22       |                 ^
May 17 09:38:22 /build/src/Common/RemoteProxyConfigurationResolver.cpp:15:41: note: the definition seen here
May 17 09:38:22    15 | std::string RemoteProxyHostFetcherImpl::fetch(const Poco::URI & endpoint, const ConnectionTimeouts & timeouts) const
May 17 09:38:22       |                                         ^
May 17 09:38:22 /build/src/Common/RemoteProxyConfigurationResolver.h:23:17: note: differing parameters are named here: ('uri'), in definition: ('endpoint')
May 17 09:38:22    23 |     std::string fetch(const Poco::URI & uri, const ConnectionTimeouts & timeouts) const override;
May 17 09:38:22       |                 ^                       ~~~
May 17 09:38:22       |                                         endpoint
May 17 09:38:22 23865 warnings generated.
May 17 09:38:22 Suppressed 24234 warnings (23864 in non-user code, 370 NOLINT).
```

File: src/Common/RemoteProxyConfigurationResolver.cpp [hunk 10]
```diff
@@ -57,50 +74,18 @@ ProxyConfiguration RemoteProxyConfigurationResolver::resolve()
...
-                    throw;
-            }
-        }
-
-        Poco::Net::HTTPResponse response;
-        auto & response_body_stream = session->receiveResponse(response);
-
-        if (response.getStatus() != Poco::Net::HTTPResponse::HTTP_OK)
-            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Proxy resolver returned not OK status: {}", response.getReason());
+        const auto proxy_host = fetcher->fetch(endpoint, timeouts);
```
Comment (CheSema) [inline src/Common/RemoteProxyConfigurationResolver.cpp:77]: Great. You deleted a bunch of  untested strange code!

**Comment by arthurpassos** [comment]
@CheSema I am returning from a work trip, will address your comments in a few days

**Comment by arthurpassos** [comment]
I don't quite understand what is happening with `binary_tidy` build, the logs show: `src/CMakeFiles/dbms.dir/Disks/ObjectStorages/ObjectStorageFactory.cpp.o -MF src/CMakeFiles/dbms.dir/Disks/ObjectStorages/ObjectStorageFactory.cpp.o.d -o src/CMakeFiles/dbms.dir/Disks/ObjectStorages/ObjectStorageFactory.cpp.o -c /build/src/Disks/ObjectStorages/ObjectStorageFactory.cpp
May 21 15:19:05 /build/src/Disks/ObjectStorages/ObjectStorageFactory.cpp:309:21: error: unused local variable 'endpoint_string' of type 'std::string' (aka 'basic_string<char, char_traits<char>, allocator<char>>') [bugprone-unused-local-non-trivial-variable,-warnings-as-errors]
May 21 15:19:05   309 |         std::string endpoint_string = endpoint.getEndpoint();
`

But I haven't changed anything in that file and I can't even find this line in the code

File: src/Common/RemoteProxyConfigurationResolver.cpp (hunk 7 above)
Comment (CheSema) [reply src/Common/RemoteProxyConfigurationResolver.cpp:26]: I think it is a little bit more tricky.

Here we are mutting all the errors from http request to the resolver. If request failed we return empty config.
https://github.com/ClickHouse/ClickHouse/blob/f82a816fdd9de00788efa15dac5b1881a8141a11/src/Common/RemoteProxyConfigurationResolver.cpp#L113-L117

I do not understand that. Why we return empty config? Why not cached config from previous resolve request? Or why not rethrow all exception up to trigger retry request?

How is it suppose to work if we have just unstable network here?

Comment (arthurpassos) [reply src/Common/RemoteProxyConfigurationResolver.cpp:26]: I keep taking this code for granted, but I agree we should let the exception propagate to hit the retry logic

**Comment by arthurpassos** [comment]
I just found out that `test_s3_storage_conf_proxy` has always been broken.

Storage conf is set up in very early stages of ClickHouse startup, and the ClickHouse instance is starting before the proxy bottle http server that responds to remote resolver requests. This causes the first request to retrieve proxy for S3 communication to fail, but it would [return empty configuration, not exception](https://github.com/ClickHouse/ClickHouse/blob/3a6f2ccf4f7dda229d668f1681e0591251e93388/src/Common/RemoteProxyConfigurationResolver.cpp#L116), thus it would bypass the proxy and reach S3. The following requests would go through the proxy because by that time the http server would already be setup, and this was filling the proxy logs, which is used to assert the proxy was actually used.

Simply re-ordering stuff does not work because the remote resolver startup script depends on a running clickhouse cluster. Another problem is that the HTTP server is not set up by docker, but rather by python scripts executed in integration tests that copy the server code over to the machine, set permissions and run. Making it even slower.

One option, tho it would reduce flexibility, is to come up with a docker-compose that maps the http server file and execute it (during `cluster.start`, all images are started before clickhouse instances, so this would work).

**Comment by arthurpassos** [comment]
I think I managed to fix it using the above approach, and it does not reduce flexibility as I predicted:

> One option, tho it would reduce flexibility, is to come up with a docker-compose that maps the http server file and execute it (during cluster.start, all images are started before clickhouse instances, so this would work).

File: src/Common/RemoteProxyConfigurationResolver.cpp [hunk 11]
```diff
@@ -16,22 +15,41 @@ namespace ErrorCodes
...
+{
+    auto request = Poco::Net::HTTPRequest(Poco::Net::HTTPRequest::HTTP_GET, endpoint.getPath(), Poco::Net::HTTPRequest::HTTP_1_1);
+    auto session = makeHTTPSession(HTTPConnectionGroupType::HTTP, endpoint, timeouts);
+
+    session->sendRequest(request);
+
+    Poco::Net::HTTPResponse response;
+    auto & response_body_stream = session->receiveResponse(response);
+
+    if (response.getStatus() != Poco::Net::HTTPResponse::HTTP_OK)
```
Comment (CheSema) [inline src/Common/RemoteProxyConfigurationResolver.cpp:28]: Could be here redirects? Do we need to follow redirects here?

Comment (arthurpassos) [reply src/Common/RemoteProxyConfigurationResolver.cpp:28]: I think it depends on the http server. Honestly, this "remote resolver" feature seems to be very unique for me and doesn't sound like a lot of people rely on it. I propose we keep it simple and implement it if the need arises.

File: src/Common/RemoteProxyConfigurationResolver.cpp [hunk 12]
```diff
@@ -16,22 +15,41 @@ namespace ErrorCodes
...
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Proxy resolver returned not OK status: {}", response.getReason());
```
Comment (CheSema) [inline src/Common/RemoteProxyConfigurationResolver.cpp:29]: `BAD_ARGUMENTS` do not understand here that error. That error refer to the user mistake, but that is out of his control.
` throw HTTPException(code, uri, status, response.getReason(), body.str());`

Comment (arthurpassos) [reply src/Common/RemoteProxyConfigurationResolver.cpp]: Is the `body` mandatory? Followed: https://github.com/ClickHouse/ClickHouse/blob/a49f5e43437f03b75a94a1af702a1862396dd568/src/IO/ReadWriteBufferFromHTTP.cpp#L405

**Comment by arthurpassos** [comment]
Broke some tests that rely on `minio`, but do not need a proxy resolver. Will work on that.

**Comment by arthurpassos** [comment]
My solution to the proxy resolver / clickhouse initialization problem broke other tests that rely on minio, but not the proxy resolver that comes bundled with it. Making the http server optional fixed some of those failures, but there are some other s3 tests that need the http server for other purposes. Making a generic setup fixed some more cases, but now there is a tricky case where a test spins up multiple http servers, not only one. https://github.com/ClickHouse/ClickHouse/blob/5a35010a8392bbc87eb5f7c5a4611281a15e3cc7/tests/integration/test_storage_s3/test.py#L729

I'll rethink this, maybe splitting the docker images is the way to go, but might introduce other problemms..

**Comment by arthurpassos** [comment]
> My solution to the proxy resolver / clickhouse initialization problem broke other tests that rely on minio, but not the proxy resolver that comes bundled with it. Making the http server optional fixed some of those failures, but there are some other s3 tests that need the http server for other purposes. Making a generic setup fixed some more cases, but now there is a tricky case where a test spins up multiple http servers, not only one.
> 
> https://github.com/ClickHouse/ClickHouse/blob/5a35010a8392bbc87eb5f7c5a4611281a15e3cc7/tests/integration/test_storage_s3/test.py#L729
> 
> I'll rethink this, maybe splitting the docker images is the way to go, but might introduce other problemms..

I thought maybe having a directory inside the test directory containing all files and loop over those on docker-compose would do the trick, but there is an extra case with http servers that live in [test/integration/helpers/s3_mocks](https://github.com/ClickHouse/ClickHouse/blob/master/tests/integration/helpers/s3_mocks/broken_s3.py)

**Comment by arthurpassos** [comment]
@CheSema Do you have an opinion on the above? I am considering to comment out the remote resolver test for S3 storage. S3 storage with proxy is already tested with proxy list and remote resolver is already tested for s3 table function and url functions.

If you have a suggestion on how to get this working, I will work on it

File: src/Common/RemoteProxyConfigurationResolver.cpp (hunk 12 above)
Comment (CheSema) [reply src/Common/RemoteProxyConfigurationResolver.cpp]: > Is the body mandatory?
No

File: tests/integration/test_s3_table_function_with_http_proxy/configs/config.d/proxy_list.xml [hunk 13]
```diff
@@ -2,7 +2,6 @@
     <proxy>
         <http>
             <uri>http://proxy1</uri>
-            <uri>http://proxy2</uri>
```
Comment (CheSema) [inline tests/integration/test_s3_table_function_with_http_proxy/configs/config.d/proxy_list.xml:5]: I do not understand why did you remove all second records?

**Comment by CheSema** [comment]
I confused with you problem with starting proxy.
Clickhouse could start with no requests to the s3-disk if it has no tables on it.
Clickhouse starts. Minio starts after. Proxy starts after minio. A test begins to run when all layers are up.

**Comment by CheSema** [comment]
> I thought maybe having a directory inside the test directory containing all files and loop over those on docker-compose would do the trick, but there is an extra case with http servers that live in [test/integration/helpers/s3_mocks](https://github.com/ClickHouse/ClickHouse/blob/master/tests/integration/helpers/s3_mocks/broken_s3.py)

We want to reuse proxys code between different tests.

**Comment by arthurpassos** [comment]
> > Aha, ClickHouse does not talk to S3, but it builds the S3Client, which depends on proxy configuration. Which means it tries to resolve the proxy configuration and the http server is not set up at that moment
> 
> It could talk to s3 but it is optional and it should not fail on error. That error is just printed to the log without any side effects, You could turn that attempt off with `skip_access_check`.
> 
> Does that error is retried?

I don't think it is retried, will check. `skip_access_check` indeed solves the issue, will revert a few commits

**Comment by arthurpassos** [comment]
> Does that error is retried?

It is not retried because it fails during [getRegionForBucket](https://github.com/ClickHouse/ClickHouse/blob/13fa946454a632c733a0a07151261b63512eabe5/src/IO/S3/Client.cpp#L730) call, which is not wrapped by the [with_retries wrapper](https://github.com/ClickHouse/ClickHouse/blob/13fa946454a632c733a0a07151261b63512eabe5/src/IO/S3/Client.cpp#L630). See [doRequest](https://github.com/ClickHouse/ClickHouse/blob/13fa946454a632c733a0a07151261b63512eabe5/src/IO/S3/Client.cpp#L539) method for more info.

**Comment by arthurpassos** [comment]
> > Does that error is retried?
> 
> It is not retried because it fails during [getRegionForBucket](https://github.com/ClickHouse/ClickHouse/blob/13fa946454a632c733a0a07151261b63512eabe5/src/IO/S3/Client.cpp#L730) call, which is not wrapped by the [with_retries wrapper](https://github.com/ClickHouse/ClickHouse/blob/13fa946454a632c733a0a07151261b63512eabe5/src/IO/S3/Client.cpp#L630). See [doRequest](https://github.com/ClickHouse/ClickHouse/blob/13fa946454a632c733a0a07151261b63512eabe5/src/IO/S3/Client.cpp#L539) method for more info.

Not only the above, but `makeHTTPSession` is throwing `ConnectionRefusedException` and `with_Retries` wrapper only catches [ConnectionResetException](https://github.com/ClickHouse/ClickHouse/blob/13fa946454a632c733a0a07151261b63512eabe5/src/IO/S3/Client.cpp#L652)

**Comment by arthurpassos** [comment]
> It is not retried because it fails during [getRegionForBucket](https://github.com/ClickHouse/ClickHouse/blob/13fa946454a632c733a0a07151261b63512eabe5/src/IO/S3/Client.cpp#L730) call, which is not wrapped by the [with_retries wrapper](https://github.com/ClickHouse/ClickHouse/blob/13fa946454a632c733a0a07151261b63512eabe5/src/IO/S3/Client.cpp#L630). See [doRequest](https://github.com/ClickHouse/ClickHouse/blob/13fa946454a632c733a0a07151261b63512eabe5/src/IO/S3/Client.cpp#L539) method for more info.

> Not only the above, but makeHTTPSession is throwing ConnectionRefusedException and with_Retries wrapper only catches [ConnectionResetException](https://github.com/ClickHouse/ClickHouse/blob/13fa946454a632c733a0a07151261b63512eabe5/src/IO/S3/Client.cpp#L652)

I have wrapped the `getRegionForBucket` call in a try-catch so it does not throw when call fails and modified retry logic so all net exceptions are retried, not only connection reset. What do you think?

**Comment by CheSema** [comment]
We did wrapper `doRequestWithRetryNetworkErrors` because AWS S3 unable to catch and retry `ConnectionResetException`. It happens when SDK reads XML. It just a bug in sdk, we mitigated that.

However, I believe AWS SDK retries `ConnectionRefusedException`. Could you show a stack? It something interesting.

**Comment by arthurpassos** [comment]
> We made wrapper `doRequestWithRetryNetworkErrors` because AWS S3 unable to catch and retry `ConnectionResetException`. It happens when SDK reads XML. It just a bug in sdk, we mitigated that.
> 
> However, I believe AWS SDK retries `ConnectionRefusedException`. Could you show a stack? It something interesting. `ConnectionRefusedException` happens pretty often, we wont survived if we did not retry that.
> 
> Even if you avoid all problems with `skip_access_check=1`. Please dig to retries of `getRegionForBucket`. It might be connected how proxy resolve is retried.

The culprit can be found in the last commit https://github.com/ClickHouse/ClickHouse/pull/63427/commits/313db20b74cb25294ea76fe4802343064a5c17e1#diff-aaad9300ade1d534bdc2dd18a47e277eefab575cc0ecbe2b7954c1e5d7133d4dL308. Proxy resolver was not being called inside the try-catch block that transforms exceptions into error return values, thus it was not being retried.

File: src/Common/RemoteProxyConfigurationResolver.cpp [hunk 14]
```diff
@@ -6,30 +6,55 @@
...
 }
 
 ProxyConfiguration RemoteProxyConfigurationResolver::resolve()
 {
     auto logger = getLogger("RemoteProxyConfigurationResolver");
 
-    auto & [endpoint, proxy_protocol, proxy_port, cache_ttl_] = remote_server_configuration;
+    auto & [endpoint, proxy_protocol_string, proxy_port, cache_ttl] = remote_server_configuration;
 
     LOG_DEBUG(logger, "Obtain proxy using resolver: {}", endpoint.toString());
```
Comment (CheSema) [inline src/Common/RemoteProxyConfigurationResolver.cpp:59]: That message would be printed for each s3 request. It is too much.
Options:
- delete it
- put it under the flag enable_s3_logging
- make logging with special logger that makes samplings

File: tests/integration/test_s3_table_function_with_http_proxy/configs/config.d/proxy_list.xml (hunk 13 above)
Comment (CheSema) [reply tests/integration/test_s3_table_function_with_http_proxy/configs/config.d/proxy_list.xml:5]: I still do not understand. May be I miss something it was a long thread.

Comment (arthurpassos) [reply tests/integration/test_s3_table_function_with_http_proxy/configs/config.d/proxy_list.xml:5]: I don't quite remember either. I am sure it was smething related to the test refactoring in this PR that was going to bit a bit harder if we were to keep multiple entries, and since it is already tested in [unit tests](https://github.com/ClickHouse/ClickHouse/blob/72372425766cec1fd713be71cf75b4b426f69244/src/Common/tests/gtest_proxy_list_configuration_resolver.cpp#L15), I thought it would be easier to just remove it.

I'll try to introduce it back.

**Comment by arthurpassos** [comment]
test_checking_s3_blobs_paranoid - https://github.com/ClickHouse/ClickHouse/issues/54882

File: tests/integration/test_s3_table_function_with_http_proxy/configs/config.d/proxy_list.xml (hunk 13 above)
Comment (arthurpassos) [reply tests/integration/test_s3_table_function_with_http_proxy/configs/config.d/proxy_list.xml]: Done

**Comment by arthurpassos** [comment]
@CheSema ping

**Review by CheSema — APPROVED** [review]
